/*
 *  Copyright (c) 2023-2024 Qualcomm Innovation Center, Inc. All rights reserved.
 *  SPDX-License-Identifier: BSD-3-Clause-Clear
 */

syntax = "proto3";

package sensorStub;
import "google/protobuf/empty.proto";
import "common_simulation.proto";
import "event_simulation.proto";

/*******************************************************************
 * Services Definition                                             *
 *******************************************************************/

service SensorFeatureManagerService {
   rpc InitService (google.protobuf.Empty) returns (GetServiceStatusReply) {}
   rpc GetFeatureList (google.protobuf.Empty) returns (GetFeatureListReply) {}
   rpc EnableFeature (SensorEnableFeature) returns (SensorFeatureManagerCommandReply) {}
   rpc DisableFeature (SensorEnableFeature) returns (SensorFeatureManagerCommandReply) {}
}

service SensorClientService {
    rpc InitService (google.protobuf.Empty) returns (GetServiceStatusReply) {}
    rpc GetSensorList (google.protobuf.Empty) returns (SensorInfoResponse) {}
    rpc Configure (google.protobuf.Empty) returns (SensorClientCommandReply) {}
    rpc GetConfiguration (google.protobuf.Empty) returns (SensorClientCommandReply) {}
    rpc GetSensorInfo (google.protobuf.Empty) returns (SensorClientCommandReply) {}
    rpc Activate (ActivateRequest) returns (SensorClientCommandReply) {}
    rpc Deactivate (DeactivateRequest) returns (SensorClientCommandReply) {}
    rpc SensorUpdateRotationMatrix (google.protobuf.Empty) returns (SensorClientCommandReply) {}
    rpc SelfTest (SelfTestRequest) returns (SelfTestResponse) {}
}

service EventDispatcherService {
    rpc isServiceAvailable(google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc registerForEvents(eventService.EventRequest) returns (stream eventService.EventResponse) {}
    rpc InjectEvent(eventService.UnsolicitedEvent) returns (google.protobuf.Empty) {}
    rpc updateFilter(eventService.EventRequest) returns (google.protobuf.Empty) {}
    rpc cleanup(eventService.CleanupRequest) returns (google.protobuf.Empty) {}
}

/*******************************************************************
 * Messages Definition                                             *
 *******************************************************************/

message GetServiceStatusReply {
    commonStub.ServiceStatus service_status = 1;
    int32 delay = 2;
}

message SensorFeatureManagerCommandReply {
    commonStub.Status status = 1;
    int32 delay = 2;
}

message SensorClientCommandReply {
    commonStub.Status status = 1;
    commonStub.ErrorCode error = 2;
    int32 delay = 3;
}

message SensorEnableFeature{
    string feature = 1;
}

message FeatureEvent{
    int32 id = 1;
    string featureName = 2;
    string events = 3;
}

message GetFeatureListReply {
    commonStub.Status status = 1;
    int32 delay = 2;
    string list = 3;
}

enum SensorType {
    ACCEL = 0;
    GYRO = 1;
}

message ActivateRequest {
    SensorType sensor_type = 1;
}

message DeactivateRequest {
    SensorType sensor_type = 1;
}

message StartReportsEvent {
    string sensor_report = 1;
}

message StreamingStoppedEvent {}

message SensorInfo {
    int32 id = 1;
    int32 sensor_type = 2;
    string name = 3;
    string vendor = 4;
    repeated float sampling_rates = 5;
    float max_sampling_rate = 6;
    uint32 max_batch_count_supported = 7;
    uint32 min_batch_count_supported = 8;
    int32 range = 9;
    int32 version = 10;
    float resolution = 11;
    float max_range = 12;
}

message SensorInfoResponse {
    repeated SensorInfo sensor_info = 1;
}

enum SelfTestType {
    SelfTest_Positive = 0;
    SelfTest_Negative = 1;
    SelfTest_All = 2;
}

enum SelfTestResult {
    Sensor_Busy = 0;
    Sensor_Idle = 1;
}

message SelfTestRequest {
    SelfTestType selftest_type = 1;
    SensorType sensor_type = 2;
}

message SelfTestResponse {
    commonStub.Status status = 1;
    commonStub.ErrorCode error = 2;
    int32 delay = 3;
    SelfTestResult selftest_result = 4;
    uint64 timestamp = 5;
}

message SelfTestFailedEvent {
    int32 sensor_mask = 1;
}